### PR TITLE
[Ready] SSthrowing + callbacks!

### LINKED ISSUE #19187: Throwing is needlessly expensive
Throwing and spacedrift excessively use sleeps, such that each item being thrown or moved sleeps then wakes up, causing a sleep overhead (about 5 to 30 proc calls, depending on how many other sleeps are in the queue) per item.

Throwing is the main issue, if there are a bunch of shit being thrown thru space, they sleep every 1ds, meaning 100 items being thrown starts to substantially eat into cpu time.

The easy fix is to make a throwing subsystem, the hard part is things that do throw_at and expect it to block so they can have code that runs after the throw finishes. Call backs could work, but it might be more feasible to just figure out workarounds around those cases.

I was going to do it, but overlay re-work needs to be finished first.

Every time the server starts lagging throw_at is rather high in the profile

Other sleep per object things could be subsystem'ed as well to avoid the needless sleeps, like orbits, space drifting, and projectiles.

### LINKED ISSUE #20305: throw_at() without diagonals_first set to TRUE is kind of wildly inaccurate
As in, you give it a target and it will totally miss that target; this happens most often with things that are really close to being at an entirely diagonal direction from the starting throw point, but aren't quite.
space station 13 2016-09-03 131157
If that bubblegum charges at that human without diagonals_first being TRUE, it'll fucking miss.

### PR DESCRIPTION
**Pull request by MrsTonedOne**
- Throwing is now a subsystem.
- It's low priority, but is a ticker subsystem so is ran before most other subsystems.
- To allow for shit to run after the throw finishes, throw_at now supports a callback. (since it can no longer block)
- **A callback datum system was created**, conversion of addtimer is planned for another PR.
- Throwing now has a limit of 512 turfs (was 600)
- Throwing now ticks every world.tick, (it converts the speed arg from turfs per 1ds to turfs per what ever tick_lag is) (under default fps, that means throwing now ticks twice as often as the old method, moving the thing half as many turfs each tick)
- Throwing now properly accounts for missed ticks.
- Throwing no longer uses sleep.
- Throwing should no longer lag since it's not filling the sleep queue up
- :new: Item throwing now imparts the momentum of the user throwing. (this is to make hitting yourself with thing you are throwing much less likely)

:cl: MrStonedOne
tweak: Throwing was refactored to cause less lag and be more precise
add: Item throwing now imparts the momentum of the user throwing. Throwing in the direction you are moving throws the item faster, throwing away from the direction you are moving throws the item slower. This should make hitting yourself with floor tiles less likely.
/:cl:

I don't know how this fits with the freeze, but i'm calling it a refactor.

fixes #19187
fixes #20305
(mostly fixes) #18670 (it can still happen in certain edge cases because movement doesn't follow normal byond ticks and can in fact happen in between them, but i'm calling that a feature since the likelihood of it happening increases the slower the item is thrown, hitting yourself with a paper when throwing it is just hilarious)

### CODE REVIEW

File: code/datums/callback.dm [hunk 1]
```diff
@@ -0,0 +1,90 @@
+/*
+	USAGE:
+
+		var/datum/callback/C = new(object|null, /proc/type/path|"procstring", arg1, arg2, ... argn)
+		var/timerid = addtimer(C, time, timertype)
+		OR
+		var/timerid = addtimer(CALLBACK(object|null, /proc/type/path|procstring, arg1, arg2, ... argn), time, timertype)
```
Comment (MrStonedOne) [inline code/datums/callback.dm]: these don't work yet since addtimer hasn't been converted, but they will.

File: code/datums/callback.dm [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+		global proc while in another global proc:
+			.procname
+			Example:
+				CALLBACK(GLOBAL_PROC, .some_proc_here)
+
+		proc defined on current(src) object (when in a /proc/ and not an override) OR overridden at src or any of it's parents:
+			.procname
+			Example:
+				CALLBACK(src, .some_proc_here)
```
Comment (MrStonedOne) [inline code/datums/callback.dm]: @Fox-McCloud this might interest you, since you were to first to sound the alarm over abusing string procs

**Comment by XDTM** [comment]
Can i run code at the end of the throw now? If so this would make tentacles work more smoothly

**Comment by MrStonedOne** [comment]
you always could. it was only throw_at_fast that didn't support that(because all it was was throw_at but didn't block). throw_at was normally blocking, as in it didn't return until the throw finished.

Now it's back to non-blocking.

Anywho, for your tentacles i wasn't sure if the addtimer was a delay, or if you thought throw_at was non-blocking, so i kept it, i'll change that over to just call the grab/stab proc directly

**Comment by XDTM** [comment]
Thanks

**Comment by PKPenguin321** [comment]
Why is MSO so based?

File: code/datums/martial/wrestling.dm [hunk 3]
```diff
@@ -182,9 +182,8 @@
 		if (T && isturf(T))
 			if (!D.stat)
 				D.emote("scream")
-			D.throw_at(T, 10, 4)
+			D.throw_at(T, 10, 4, callback = CALLBACK(D, /mob/living/carbon/human/.Weaken, 2))
 			D.Weaken(2)
```
Comment (optimumtact) [inline code/datums/martial/wrestling.dm]: won't this weaken them, then weaken them again?

Comment (MrStonedOne) [inline code/datums/martial/wrestling.dm]: ya, i forgot to remove the line once i made it a callback

**Comment by Iamgoofball** [comment]
What's a callback and how does it work in mememan's terms

**Comment by RemieRichards** [comment]
@Iamgoofball the name is the clue, "callback"

"Can I give you a call back in 5 minutes?"
"Sure"

is the same as
Procname() Called
Procname() Running
Procname() Finishes
Callbackname() Called automatically

**Comment by MrStonedOne** [comment]
http://www.byond.com/forum/?post=2191707 This is one of those things that's easier to explain with code then with words. it's just a wrapper around procs that take procs to call as arguments. the thingtocall, proctocall, args.... mess that addtimer has, that sorttim has, etc. this is a generic wrapper around that concept.

Lets say you want `/mob/living/proc/after_item_throw` to be called after a throw mode throw finishes, Since throwing is a subsystem, you can't just call `throw_at`, then call `after_item_throw`, so you have to tell `throw_at` to call it when the throw finishes. with this system, you just do:
````
/mob/living/proc/click_on_tile(some shit here)
	if (throwmode)
		item.throw_at(thing_clicked,some_other_shit, callback = CALLBACK(src, .proc/after_item_throw, item, some, other, args, to, call, after_item_throw, with.)
````

Now, all `throw_at` has to do, is store the created callback datum, and run `callback`.`Invoke()` when the throw finishes to run after_item_throw

a callback is just a reference to a proc call and args that's easier to use and work with.

**Comment by Iamgoofball** [comment]
💯🍆👌👌👌👌👌👌👌👌👌👌👌💯💯💯<😘😘

**Comment by Iamgoofball** [comment]
I wonder if we can use this for some Tesla stuff

**Comment by RemieRichards** [comment]
@MrStonedOne 

>easier in code than in words
>literally explained in 2 words

uh huh.

**Comment by MrStonedOne** [comment]
>I wonder if we can use this for some Tesla stuff

Your mind is weird conceptually.

Explain to me how you arrived at that conclusion. where you read about that, and thought tesla.

**Comment by Iamgoofball** [comment]
right so we might be able to abuse callbacks to clean up tesla_zap, I'll explain better when I'm at a computer

**Comment by optimumtact** [comment]
some wack ass conflitcs going on here

**Comment by MrStonedOne** [comment]
Test merging on bagil

**Comment by AnturK** [comment]
Is this finished ?

**Comment by MrStonedOne** [comment]
oh. uhh, I was trying to fix the issue where throwing while running caused you to hit yourself. i'll go re-checkout that branch and see where i was on that.

**Comment by MrStonedOne** [comment]
I'm gonna rebase then re-check for anything merged recently that depended on old behavior, then this can be merged.

**Pull request merged**